src/components/plots/ViolinPlot.jsx

```diff
@@ -0,0 +1,153 @@
+import React, { useState, useEffect } from 'react';
+import { useSelector, useDispatch } from 'react-redux';
+import PropTypes from 'prop-types';
+import { Vega } from 'react-vega';
+
+import { Skeleton } from 'antd';
+import PlatformError from '../PlatformError';
+import { generateSpec, generateData } from '../../utils/plotSpecs/generateViolinSpec';
+import { loadGeneExpression, loadPaginatedGeneProperties } from '../../redux/actions/genes';
+import { loadCellSets } from '../../redux/actions/cellSets';
+import { updatePlotConfig } from '../../redux/actions/componentConfig/index';
+
+const ViolinPlot = (props) => {
+  const {
+    experimentId, config, plotUuid, plotData, actions,
+  } = props;
+  const dispatch = useDispatch();
+
+  const geneExpression = useSelector((state) => state.genes.expression);
+  const cellSets = useSelector((state) => state.cellSets);
+  const [plotSpec, setPlotSpec] = useState({});
+  const highestDispersionLoading = useSelector(
+    (state) => state.genes.properties.views[plotUuid]?.fetching,
+  );
+  const highestDispersionError = useSelector(
+    (state) => state.genes.properties.views[plotUuid]?.error,
+  );
+  const highestDispersionGene = useSelector(
+    (state) => (state.genes.properties.views[plotUuid]?.data
+      ? state.genes.properties.views[plotUuid]?.data[0] : undefined),
+  );
+  const PROPERTIES = ['dispersions'];
+  const tableState = {
+    pagination: {
+      current: 1, pageSize: 1, showSizeChanger: true, total: 0,
+    },
+    geneNamesFilter: null,
+    sorter: { field: PROPERTIES[0], columnKey: PROPERTIES[0], order: 'descend' },
+  };
+
+  if (config?.shownGene === 'notSelected' && !highestDispersionLoading && !highestDispersionGene) {
+    dispatch(loadPaginatedGeneProperties(experimentId, PROPERTIES, plotUuid, tableState));
+  }
+  useEffect(() => {
+    if (cellSets.loading && !cellSets.error) {
+      dispatch(loadCellSets(experimentId));
+    }
+  }, [experimentId]);
+  useEffect(() => {
+    if (config?.shownGene === 'notSelected' && highestDispersionGene) {
+      dispatch(updatePlotConfig(plotUuid, { shownGene: highestDispersionGene }));
+      dispatch(loadGeneExpression(experimentId, [highestDispersionGene], plotUuid));
+    }
+    if (config?.shownGene !== 'notSelected' && config) {
+      dispatch(loadGeneExpression(experimentId, [config.shownGene], plotUuid));
+    }
+  }, [highestDispersionGene, config?.shownGene]);
+
+  useEffect(() => {
+    if (plotData) {
+      setPlotSpec(generateSpec(config, plotData));
+      return;
+    }
+
+    if (config
+      && Object.getOwnPropertyDescriptor(geneExpression.data, config.shownGene)
+      && !geneExpression.error
+      && !cellSets.loading
+      && !cellSets.error) {
+      const expressionType = config.normalised === 'normalised' ? 'expression' : 'raw TO-DO';
```

Comment:
is it TODO though? Did we not do it yet? I am pretty sure the heatmap in data exploration can show two versions of gene expression. Can you check?

Reply:
Yes. Heatmap has `zOrder`. That's different from raw: https://this-is-biomage.slack.com/archives/C01CP46TJ0Z/p1618987823110200

Reply:
fair enough, we will have to do it as part of another work